### PR TITLE
Auto Annotations: Backend

### PR DESCRIPTION
**Pull request by guynir42**
Attempts to fix issue #961 

I have added two fields to the comment model:
-is_auto which is true/false
-altdata which is a parsed JSONB field that can contain all sorts of data. 

The idea is that when you input "z=0.5" that would be registered into altdata and so can be searchable. 

The is_auto needs to be specified by any API call from any robots making comments (I still have not found where such comments are made). 

I've also added a simple parser that handles human comments given as keyword=value to also be added into altdata even when is_auto is false. 

I don't think this is done but it is a starting point so please tell me what else we need to do. 

Also, I am not sure but I think this requires some changes to the database.

### CODE REVIEW

File: skyportal/handlers/api/comment.py
Comment (dannygoldstein) [inline skyportal/handlers/api/comment.py:234]: I am not sure we want to be getting into the business of writing API middleware to parse user comments like this. The logic here is already somewhat complicated (strip this, lower case that, check the data type, etc. -- many places for subtle issues to crop up when dealing with real-world user data). Once we set the precedent of doing this, we open ourselves up to having to write all kinds of custom parsers, each one adding a new layer of complexity to the backend. I say we should let the data types speak contractually for themselves, and, to the extent possible, not deviate from CRUD.

File: skyportal/handlers/api/comment.py
Comment (dannygoldstein) [inline skyportal/handlers/api/comment.py:75]: `altdata` and `text` creep into each other here somewhat. The content of `text` influences the content of `altdata`, and the content of `altdata` is handled different depending on whether `is_auto` is flagged. I think this is a sign that a separate database model is needed for autoannotations.  To the extent possible, I would like to keep our API for individual database classes as close to basic CRUD operations (https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) as possible. This makes the API contract presented to the user extremely clear, improves the long-term maintainability of the code, and makes it easier to extend the data model later on.

**Review by dannygoldstein — CHANGES_REQUESTED** [review]
I worry that the data model here overloads the `Comment` object beyond what it was designed to accommodate. I think it would be simplest to break out the notion of arbitrary JSON user data to an `Annotation` object that has its own handler and frontend components (i.e., displayed in a materialui-datatable, instead of a CommentList).

**Comment by guynir42** [comment]
I did a lot of copy-paste for this. I'm not sure if this is better than combining the comments and annotations, or at least having annotations inherit from comments. 

Either way, I couldn't find where to get a new ACL for annotations, so I left it the same ACL as comments.

**Comment by dannygoldstein** [comment]
@guynir42 Looks like maybe something is not being imported, causing the docs build to fail?

File: skyportal/handlers/api/annotation.py
Comment (dannygoldstein) [inline skyportal/handlers/api/annotation.py:153]:
```suggestion
```

We dont want to handle redshifts here. There is going to be a separate handler for redshifts based on the classifications API

File: skyportal/handlers/api/annotation.py
Comment (dannygoldstein) [inline skyportal/handlers/api/annotation.py:163]:
```suggestion
    @permissions(['Annotate'])
```

File: skyportal/model_util.py
Comment (dannygoldstein) [inline skyportal/model_util.py:9]:
```suggestion
    'Annotate',
```

File: skyportal/models.py
Comment (dannygoldstein) [inline skyportal/models.py:1425]:
```suggestion
    origin = sa.Column(sa.String, doc='Annotation origin.', index=True, nullable=False)
```

Comment (dannygoldstein) [reply skyportal/models.py:1425]: Can you expand the docs here somewhat?

File: skyportal/handlers/api/annotation.py
Comment (dannygoldstein) [inline skyportal/handlers/api/annotation.py:90]: This needs to be validated

File: skyportal/tests/api/test_annotations.py
Comment (dannygoldstein) [inline skyportal/tests/api/test_annotations.py:20]: This will need to be changed when the validation is handled via a Marshmallow schema

File: skyportal/tests/api/test_annotations.py
Comment (dannygoldstein) [inline skyportal/tests/api/test_annotations.py:28]: The empty string should be allowed as a valid origin. It should only barf when origin is null

File: skyportal/tests/api/test_annotations.py
Comment (dannygoldstein) [inline skyportal/tests/api/test_annotations.py:48]: Defer redshift handling to redshift handler - to be implemented cc @acrellin

File: skyportal/tests/api/test_annotations.py
Comment (dannygoldstein) [inline skyportal/tests/api/test_annotations.py:71]: I would refrain from checking the messages as this may increase the brittleness of the test suite but I defer to @stefanv  on this if he thinks otherwise

File: skyportal/tests/api/test_annotations.py
Comment (dannygoldstein) [inline skyportal/tests/api/test_annotations.py:146]:
```suggestion
    assert data['data']['data'] == {'sdss_redshift': 0.15}
```
Suggest replacing this throughout the test suite with `sdss_redshift` or similar to avoid any confusion about how and where redshifts are set

**Review by dannygoldstein — CHANGES_REQUESTED** [review]
Few changes requested below, thanks for the thorough test suite.

File: skyportal/handlers/api/annotation.py
Comment (dannygoldstein) [reply skyportal/handlers/api/annotation.py:90]: I.e., need to validate that `data.get('data')` returns valid JSON

File: skyportal/handlers/api/annotation.py
Comment (guynir42) [reply skyportal/handlers/api/annotation.py:88]: Ok I will try to find some examples.

File: skyportal/handlers/api/annotation.py
Comment (guynir42) [reply skyportal/handlers/api/annotation.py:153]: I know, but this functionality exists in the current Comment code, so I thought it would be good to pull it out only after we have something else instead.

File: skyportal/models.py
Comment (guynir42) [reply skyportal/models.py:1425]: yes, good idea.

File: skyportal/handlers/api/annotation.py
Comment (guynir42) [reply skyportal/handlers/api/annotation.py:90]: I need to check that I get a python dictionary?

File: skyportal/tests/api/test_annotations.py
Comment (guynir42) [reply skyportal/tests/api/test_annotations.py:28]: I'm not sure about this. If people are allowed to enter empty origin it could cause a lot of trouble... Part of the whole annotations API is that you MUST give an origin when posting these. If '' is legal how is that enforced?

File: skyportal/tests/api/test_annotations.py
Comment (guynir42) [reply skyportal/tests/api/test_annotations.py:146]: I would prefer to use more realistic examples, both for the data and the origin. Can you think of something more concrete than "kowalski" and "redshift"?

File: skyportal/handlers/api/annotation.py
Comment (dannygoldstein) [reply skyportal/handlers/api/annotation.py:90]: not just any python dictionary, but valid JSON. if you use a marshmallow schema this validation will be handled automatically for you.

File: skyportal/tests/api/test_annotations.py
Comment (dannygoldstein) [reply skyportal/tests/api/test_annotations.py:28]: `''` is not an empty origin. It is the empty string, which is a valid string. If you want to require that an origin string consist of at least 1 alphanumeric character then that is not what is being validated in the current handler. For example, I could post `' '` and the handler as written would accept that as valid user input. That is fine. As long as it is a unique string, that is OK. We should, to the extent possible, let data types specify valid contracts with the users.

File: skyportal/tests/api/test_annotations.py
Comment (dannygoldstein) [reply skyportal/tests/api/test_annotations.py:146]: Kowalski is fine, and I would use something like `offset_from_host_galaxy`

File: skyportal/tests/api/test_annotations.py
Comment (guynir42) [reply skyportal/tests/api/test_annotations.py:28]: I see your point. But if we don't limit anything (maybe 1+ alphanumeric is good) then it leaves the option for abusing the system. If I'm writing a bot and I don't have patience to read the definition of "origin" I could just leave it empty. Or put a " " or a "-" as input until it works and I wouldn't care if that is actually not informative. 

I think we should enforce some reasonable constraint on this. It is easier than "educating" the users to be good and properly fill out their origin...

File: skyportal/handlers/api/annotation.py
Comment (guynir42) [reply skyportal/handlers/api/annotation.py:88]: I added a Annotations.__schema__() and used it to validate inputs but now it complains that it doesn't get an author_id and that it doesn't recognize group_ids.

File: skyportal/handlers/api/annotation.py
Comment (dannygoldstein) [inline skyportal/handlers/api/annotation.py:99]: Excellent

File: skyportal/models.py
Comment (dannygoldstein) [inline skyportal/models.py:1465]: Move to top of class

**Review by dannygoldstein — APPROVED** [review]
Excellent work @guynir42

File: skyportal/tests/api/test_annotations.py
Comment (kmshin1397) [inline skyportal/tests/api/test_annotations.py:4]: I love how thorough the tests are for the annotations. Great work!

**Review by kmshin1397 — APPROVED** [review]
@guynir42 I have one very small change (just copy-paste to duplicate a recent change to Comments) but otherwise looks great to me! Ping me on Slack when you get that in and I'll merge this.

File: skyportal/handlers/api/annotation.py
Comment (guynir42) [reply skyportal/handlers/api/annotation.py]: OK I just uploaded it. You can make sure it is ok and merge.

**Review by kmshin1397 — APPROVED** [review]
Great, will merge when green

**Pull request merged**
